AX: Implement NextSentenceEnd/PrevSentenceStart APIs off of the main thread

hoffmanjoshua opened the pull request:
#### ff546a4b6c04ed6c47dc1851f3f1268ac7e99a85
<pre>
AX: Implement NextSentenceEnd/PrevSentenceStart APIs off of the main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=281541">https://bugs.webkit.org/show_bug.cgi?id=281541</a>
<a href="https://rdar.apple.com/138005738">rdar://138005738</a>

Reviewed by Tyler Wilcock and Andres Gonzalez.

Moving NextSentenceEnd and PreviousSentenceStart off of the main thread allows the rest
of text-marker-sentence-nav.html to be ported to ax-thread-text-apis/. This patch updates
how we handle `AXTextUnit::Sentence` to appropriately handle these APIs and match the
existing live tree behavior. This test fully passes when run with AX_THREAD_TEXT_APIS
enabled.

This also includes build fixes after 285102@main (Port AXID to ObjectIdentifier).

* LayoutTests/accessibility/ax-thread-text-apis/text-marker-sentence-nav-expected.txt:
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-sentence-nav.html:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::characterRangeForLine const):
(WebCore::AXTextMarker::lineNumberForIndex const):
(WebCore::AXTextMarker::findLastBefore const):
(WebCore::previousSentenceStartFromOffset):
(WebCore::nextSentenceEndFromOffset):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::sentenceRange const):
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::textMarkerRange const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerForTextMarker:atUnit:]):

Canonical link: <a href="https://commits.webkit.org/285413@main">https://commits.webkit.org/285413@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/412b8d6bce86c545317a7be0043d044c73156b61

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72355 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51776 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76533 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23564 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74470 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59580 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23386 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57001 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15517 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75422 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46887 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62311 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37436 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43540 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19782 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21914 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65430 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20139 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78203 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16596 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19279 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65456 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16644 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62328 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64724 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16022 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12979 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6615 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47574 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2358 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48643 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49931 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48386 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -897,7 +911,9 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
-                if (currentRuns->containingBlock != currentObject->textRuns()->containingBlock || lastObjectIsEditable != !!currentObject->editableAncestor() || (currentObject->roleValue() == AccessibilityRole::LineBreak && !currentObject->editableAncestor()))
+                // For sentences, don't stop at line breaks, since the text break iterator needs to find the next sentence boundary, which isn't necessarily at a break.
+                bool shouldStopAtLineBreaks = textUnit == AXTextUnit::Word && currentObject->roleValue() == AccessibilityRole::LineBreak && !currentObject->editableAncestor();
+                if (currentRuns->containingBlock != currentObject->textRuns()->containingBlock || lastObjectIsEditable != !!currentObject->editableAncestor() || shouldStopAtLineBreaks)
```
Should we put the `shouldStopAtLineBreaks` check before the `lastObjectIsEditable != !!currentObject->editableAncestor()` check since it's quite a bit cheaper?

hoffmanjoshua replied:
Yeah good call, will move that up front!

AndresGonzalezApple commented on LayoutTests/accessibility/ax-thread-text-apis/text-marker-sentence-nav.html:
```diff
@@ -107,10 +105,35 @@
+        output += `Pre sentence start to next sentence end: ${preAndNextSentence}\n`;
     }
+
+    function verifySentences(obj, startMarker, sentenceCount) {
```
obj -> axElement

AndresGonzalezApple commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -750,7 +750,18 @@ static int previousSentenceStartFromPosition(StringView text, unsigned position)
+    if (!text.substring(position, endPosition).containsOnly<isASCIIWhitespace>()) {
+        // To match AXObjectCache::nextBoundary, don't include a newline character at the end of sentences.
+        while (endPosition > 0 && text.length() && text.substring(0, endPosition).endsWith('\n'))
+            endPosition -= 1;
```
--endPosition

AndresGonzalezApple commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -750,7 +750,18 @@ static int previousSentenceStartFromPosition(StringView text, unsigned position)
+            endPosition += 1;
+    }
+    return endPosition;
 }
```
I would rather use offset or index instead of position for the naming here, since "position" is used in this context for instances of Position and VisiblePosition classes.

AndresGonzalezApple commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -880,10 +891,13 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
                 if (direction == AXDirection::Previous && end <= objectBorder && end != -1)
                     resultMarker = AXTextMarker(*currentObject, end);
-                else if (direction == AXDirection::Next && end != -1 && end >= objectBorder)
+                else if (direction == AXDirection::Next && end != -1 && end >= objectBorder && end != (int)offset) {
```
Is there a safer way to compare than C-casting an unsigned to an int?

hoffmanjoshua replied:
Thought about this too, and our offset would have to be greater than the max int size to overflow (which would mean an offset of over 2 billion). Happy to make this safer though. Do you have any suggestions?

AndresGonzalezApple replied:
I believe there is a safe way of doing this in WTF but don't remember exactly from top of my head.

hoffmanjoshua replied:
Updated the PR to do the casting—using `Checked<Int>` to cast will automatically handle overflows.

AndresGonzalezApple commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -1032,11 +1048,11 @@ AXTextMarkerRange AXTextMarker::sentenceRange(SentenceRangeType type) const
     if (type == SentenceRangeType::Current) {
         startMarker = previousSentenceStart();
-        endMarker = nextSentenceEnd();
+        endMarker = startMarker.nextSentenceEnd();
```
Not in this change, but think we should have these functions, nextSentenceEnd() and similar, as standalone functions and not as AXTextMarker's methods.

hoffmanjoshua replied:
To me, these make sense as part of the AXTextMarker class, since they are types of movements you would want to make from an individual text marker. Are you suggesting that these become static on AXTextMarker, or completely independent?

AndresGonzalezApple replied:
It's a matter of design, but I'd say these are operations you perform on a TextMarker, involving more knowledge that what a TextMarker has. So, I'd say they should be functions outside the class. But it is an arguable point.

hoffmanjoshua committed:
AX: Implement NextSentenceEnd/PrevSentenceStart APIs off of the main thread
https://bugs.webkit.org/show_bug.cgi?id=281541
rdar://138005738

Reviewed by Tyler Wilcock and Andres Gonzalez.

Moving NextSentenceEnd and PreviousSentenceStart off of the main thread allows the rest
of text-marker-sentence-nav.html to be ported to ax-thread-text-apis/. This patch updates
how we handle `AXTextUnit::Sentence` to appropriately handle these APIs and match the
existing live tree behavior. This test fully passes when run with AX_THREAD_TEXT_APIS
enabled.

This also includes build fixes after 285102@main (Port AXID to ObjectIdentifier).

* LayoutTests/accessibility/ax-thread-text-apis/text-marker-sentence-nav-expected.txt:
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-sentence-nav.html:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::characterRangeForLine const):
(WebCore::AXTextMarker::lineNumberForIndex const):
(WebCore::AXTextMarker::findLastBefore const):
(WebCore::previousSentenceStartFromOffset):
(WebCore::nextSentenceEndFromOffset):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::sentenceRange const):
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::textMarkerRange const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerForTextMarker:atUnit:]):

Canonical link: https://commits.webkit.org/285413@main
